versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java

```diff
@@ -1233,4 +1261,92 @@ public void nullHandlingInArrays() throws ObjTooLargeException, ObjNotFoundExcep
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v1);
+    // exists, same version - delete must succeed
+    soft.assertThat(persist.deleteConditional(v1)).isTrue();
+    long t = System.currentTimeMillis();
+
+    soft.assertThatThrownBy(() -> persist.fetchObj(v1.id()))
+        .isInstanceOf(ObjNotFoundException.class);
+
+    if ("Cassandra".equals(persist.name())) {
+      // MUST sleep here, otherwise the tombstone's timestamp might be equal to the INSERT's
+      // timestamp of the storeObj() below, which would wrongly shadow the write's timestamp.
+      long sleepMillis = t + 2 - System.currentTimeMillis();
+      if (sleepMillis > 0) {
+        Thread.sleep(sleepMillis);
+      }
+    }
+
+    // can store again
+    soft.assertThat(persist.storeObj(v1)).isTrue();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v1);
+  }
+
+  @Test
+  public void conditionalUpdate() throws Exception {
+    VersionedTestObj v1 =
+        VersionedTestObj.builder().id(randomObjId()).someValue("initial").versionToken("1").build();
+    VersionedTestObj v2 =
+        VersionedTestObj.builder().from(v1).someValue("version 2").versionToken("2").build();
+    VersionedTestObj v3 =
+        VersionedTestObj.builder().from(v1).someValue("version 3").versionToken("3").build();
+
+    // same version - must throw IAE
+    soft.assertThatIllegalArgumentException().isThrownBy(() -> persist.updateConditional(v1, v1));
+    // different IDs - must throw IAE
+    soft.assertThatIllegalArgumentException()
+        .isThrownBy(
+            () ->
+                persist.updateConditional(
+                    v1, VersionedTestObj.builder().from(v1).id(randomObjId()).build()));
+    // different types - must throw IAE
+    soft.assertThatIllegalArgumentException()
+        .isThrownBy(
+            () ->
+                persist.updateConditional(
+                    v1,
+                    VersionedTestObj2.builder()
+                        .id(v1.id())
+                        .versionToken("x")
+                        .otherValue("blah")
+                        .build()));
+
+    soft.assertThat(persist.storeObj(v1)).isTrue();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v1);
+
+    // exists, but different version - update must not succeed
+    soft.assertThat(persist.updateConditional(v3, v2)).isFalse();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v1);
+
+    // exists, expected version - update must succeed
+    soft.assertThat(persist.updateConditional(v1, v2)).isTrue();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v2);
+
+    // exists, previous expected version - update must not succeed
+    soft.assertThat(persist.updateConditional(v1, v2)).isFalse();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v2);
+
+    // exists, expected version - update must succeed
+    soft.assertThat(persist.updateConditional(v2, v3)).isTrue();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v3);
```

Comment:
Maybe test here the corner case where a versioned object exists, but with wrong type:

```java
    persist.deleteObj(v1.id());
    VersionedTestObj2 wrong = VersionedTestObj2.builder().from(v1).otherValue("blah").build();
    persist.storeObj(wrong);
    soft.assertThat(persist.updateConditional(v1, v2)).isFalse();
```

Reply:
Good idea, added